Adapt test ensemble to support sklearn v0.24.0

* xclim version: master

### Description

`scikit-learn` released a new minor version (https://pypi.org/project/scikit-learn/) over the holidays, and it's breaking tests.


### What I Did
<!--Paste the command(s) you ran and the output.
If there was a crash, please include the traceback below.-->
```
$ tox
```

### What I Received
<!--Paste the output or the stack trace of the problem you experienced here.-->
```
xclim/testing/tests/test_ensembles.py:215 (TestEnsembleReduction.test_kmeans_rsqcutoff)
[0, 1, 3, 4, 6, 7, 8, 10, 11, 15, 18, 20, 22] != [0, 1, 3, 4, 6, 7, 8, 9, 10, 11, 12, 15, 20, 22]

Expected :[0, 1, 3, 4, 6, 7, 8, 9, 10, 11, 12, 15, 20, 22]
Actual   :[0, 1, 3, 4, 6, 7, 8, 10, 11, 15, 18, 20, 22]
<Click to see difference>

self = <xclim.testing.tests.test_ensembles.TestEnsembleReduction object at 0x7fcc8c144910>

    def test_kmeans_rsqcutoff(self):
        pytest.importorskip("sklearn", minversion="0.22")
        ds = open_dataset(self.nc_file)
    
        # use random state variable to ensure consistent clustering in tests:
        [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
            data=ds.data, method={"rsq_cutoff": 0.9}, random_state=42, make_graph=False
        )
    
>       assert ids == [0, 1, 3, 4, 6, 7, 8, 9, 10, 11, 12, 15, 20, 22]
E       assert [0, 1, 3, 4, 6, 7, 8, 10, 11, 15, 18, 20, 22] == [0, 1, 3, 4, 6, 7, 8, 9, 10, 11, 12, 15, 20, 22]

xclim/testing/tests/test_ensembles.py:225: AssertionError
...
xclim/testing/tests/test_ensembles.py:238 (TestEnsembleReduction.test_kmeans_rsqopt)
[3, 4, 5, 7, 10, 11, 12, 13] != [4, 5, 7, 10, 11, 12, 13]

Expected :[4, 5, 7, 10, 11, 12, 13]
Actual   :[3, 4, 5, 7, 10, 11, 12, 13]
<Click to see difference>

self = <xclim.testing.tests.test_ensembles.TestEnsembleReduction object at 0x7fcc857a5940>

    def test_kmeans_rsqopt(self):
        pytest.importorskip("sklearn", minversion="0.22")
        ds = open_dataset(self.nc_file)
        [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
            data=ds.data,
            method={"rsq_optimize": None},
            random_state=42,
            make_graph=False,
        )
>       assert ids == [4, 5, 7, 10, 11, 12, 13]
E       assert [3, 4, 5, 7, 10, 11, 12, 13] == [4, 5, 7, 10, 11, 12, 13]

xclim/testing/tests/test_ensembles.py:248: AssertionError
...
xclim/testing/tests/test_ensembles.py:311 (TestEnsembleReduction.test_kmeans_variweights)
self = <xclim.testing.tests.test_ensembles.TestEnsembleReduction object at 0x7fcc8c14b790>

    def test_kmeans_variweights(self):
        pytest.importorskip("sklearn", minversion="0.22")
        ds = open_dataset(self.nc_file)
        # Test sample weights
        var_weights = np.ones(ds.data.shape[1])
        # reduce weights for some variables
        var_weights[3:] = 0.25
    
        [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
            data=ds.data,
            method={"rsq_cutoff": 0.9},
            random_state=42,
            make_graph=False,
            variable_weights=var_weights,
        )
        assert ids == [1, 3, 8, 10, 13, 14, 16, 19, 20]
        assert len(ids) == 9
    
        # using RSQ optimize
        [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
            data=ds.data,
            method={"rsq_optimize": None},
            random_state=42,
            make_graph=False,
            variable_weights=var_weights,
        )
    
        assert ids == [2, 4, 8, 13, 14, 22]
        assert len(ids) == 6
    
        # try zero weights
        var_weights = np.ones(ds.data.shape[1])
        var_weights[[1, 4]] = 0
    
        [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
            data=ds.data,
            method={"rsq_optimize": None},
            random_state=42,
            make_graph=False,
            variable_weights=var_weights,
        )
        # Results here may change according to sklearn version, hence the *isin* intead of ==
>       assert all(np.isin([4, 12, 13, 16], ids))
E       assert False
E        +  where False = all(array([False,  True,  True,  True]))
E        +    where array([False,  True,  True,  True]) = <function isin at 0x7fccbdf04430>([4, 12, 13, 16], [8, 11, 12, 13, 16, 20])
E        +      where <function isin at 0x7fccbdf04430> = np.isin

xclim/testing/tests/test_ensembles.py:354: AssertionError
```

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -28,7 +28,7 @@ Internal changes
 ~~~~~~~~~~~~~~~~
 * `core.cfchecks.check_valid` now accepts a sequence of strings as its `expected` argument.
 * Clean up in the tests to speed up testing. Addition of a marker to exclude slow tests when needed.
-
+* Fixes in the tests to support sklearn v0.24, clisops v0.5 and build xarray@master againt python 3.7.
 
 0.22.0 (2020-12-07)
 -------------------

--- a/setup.cfg
+++ b/setup.cfg
@@ -1,5 +1,5 @@
 [bumpversion]
-current_version = 0.22.2-beta
+current_version = 0.22.3-beta
 commit = False
 tag = False
 parse = (?P<major>\d+)\.(?P<minor>\d+).(?P<patch>\d+)(\-(?P<release>[a-z]+))?

--- a/setup.py
+++ b/setup.py
@@ -9,7 +9,7 @@ URL = "https://github.com/Ouranosinc/xclim"
 AUTHOR = "Travis Logan"
 AUTHOR_EMAIL = "logan.travis@ouranos.ca"
 REQUIRES_PYTHON = ">=3.6.0"
-VERSION = "0.22.2-beta"
+VERSION = "0.22.3-beta"
 LICENSE = "Apache Software License 2.0"
 
 with open("README.rst") as readme_file:

--- a/tests/test_ensembles.py
+++ b/tests/test_ensembles.py
@@ -214,30 +214,30 @@ class TestEnsembleReduction:
     nc_file = os.path.join("EnsembleReduce", "TestEnsReduceCriteria.nc")
 
     def test_kmeans_rsqcutoff(self):
-        pytest.importorskip("sklearn", minversion="0.22")
+        pytest.importorskip("sklearn", minversion="0.24.1")
         ds = open_dataset(self.nc_file)
 
         # use random state variable to ensure consistent clustering in tests:
         [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
-            data=ds.data, method={"rsq_cutoff": 0.9}, random_state=42, make_graph=False
+            data=ds.data, method={"rsq_cutoff": 0.5}, random_state=42, make_graph=False
         )
 
-        assert ids == [0, 1, 3, 4, 6, 7, 8, 9, 10, 11, 12, 15, 20, 22]
-        assert len(ids) == 14
+        assert ids == [4, 7, 10, 23]
+        assert len(ids) == 4
 
         # Test max cluster option
         [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
             data=ds.data,
-            method={"rsq_cutoff": 0.9},
+            method={"rsq_cutoff": 0.5},
             random_state=42,
             make_graph=False,
-            max_clusters=10,
+            max_clusters=3,
         )
-        assert ids == [0, 1, 3, 4, 6, 7, 10, 11, 18, 20]
-        assert len(ids) == 10
+        assert ids == [4, 7, 23]
+        assert len(ids) == 3
 
     def test_kmeans_rsqopt(self):
-        pytest.importorskip("sklearn", minversion="0.22")
+        pytest.importorskip("sklearn", minversion="0.24.1")
         ds = open_dataset(self.nc_file)
         [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
             data=ds.data,
@@ -245,8 +245,8 @@ class TestEnsembleReduction:
             random_state=42,
             make_graph=False,
         )
-        assert ids == [4, 5, 7, 10, 11, 12, 13]
-        assert len(ids) == 7
+        assert ids == [3, 4, 5, 7, 10, 11, 12, 13]
+        assert len(ids) == 8
 
     def test_kmeans_nclust(self):
         ds = open_dataset(self.nc_file)
@@ -272,13 +272,13 @@ class TestEnsembleReduction:
 
         [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
             data=ds.data,
-            method={"rsq_cutoff": 0.9},
+            method={"rsq_cutoff": 0.5},
             random_state=42,
             make_graph=False,
             sample_weights=sample_weights,
         )
-        assert ids == [0, 1, 3, 4, 5, 6, 7, 10, 11, 18, 20]
-        assert len(ids) == 11
+        assert ids == [0, 20, 23]
+        assert len(ids) == 3
 
         # RSQ optimize
         sample_weights = np.ones(ds.data.shape[0])
@@ -310,7 +310,7 @@ class TestEnsembleReduction:
         assert len(ids) == 6
 
     def test_kmeans_variweights(self):
-        pytest.importorskip("sklearn", minversion="0.22")
+        pytest.importorskip("sklearn", minversion="0.24.1")
         ds = open_dataset(self.nc_file)
         # Test sample weights
         var_weights = np.ones(ds.data.shape[1])
@@ -351,8 +351,8 @@ class TestEnsembleReduction:
             variable_weights=var_weights,
         )
         # Results here may change according to sklearn version, hence the *isin* intead of ==
-        assert all(np.isin([4, 12, 13, 16], ids))
-        assert len(ids) == 5
+        assert all(np.isin([12, 13, 16], ids))
+        assert len(ids) == 6
 
     def test_kmeans_modelweights(self):
         ds = open_dataset(self.nc_file)
@@ -392,27 +392,27 @@ class TestEnsembleReduction:
         "matplotlib.pyplot" not in sys.modules, reason="matplotlib.pyplot is required"
     )
     def test_kmeans_rsqcutoff_with_graphs(self):
-        pytest.importorskip("sklearn", minversion="0.22")
+        pytest.importorskip("sklearn", minversion="0.24.1")
         ds = open_dataset(self.nc_file)
 
         # use random state variable to ensure consistent clustering in tests:
         [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
-            data=ds.data, method={"rsq_cutoff": 0.9}, random_state=42, make_graph=True
+            data=ds.data, method={"rsq_cutoff": 0.5}, random_state=42, make_graph=True
         )
 
-        assert ids == [0, 1, 3, 4, 6, 7, 8, 9, 10, 11, 12, 15, 20, 22]
-        assert len(ids) == 14
+        assert ids == [4, 7, 10, 23]
+        assert len(ids) == 4
 
         # Test max cluster option
         [ids, cluster, fig_data] = ensembles.kmeans_reduce_ensemble(
             data=ds.data,
-            method={"rsq_cutoff": 0.9},
+            method={"rsq_cutoff": 0.5},
             random_state=42,
             make_graph=True,
-            max_clusters=10,
+            max_clusters=3,
         )
-        assert ids == [0, 1, 3, 4, 6, 7, 10, 11, 18, 20]
-        assert len(ids) == 10
+        assert ids == [4, 7, 23]
+        assert len(ids) == 3
 
     @pytest.mark.parametrize(
         "crit,num_select,expected",

--- a/tests/test_subset.py
+++ b/tests/test_subset.py
@@ -662,8 +662,7 @@ class TestSubsetShape:
                     sub[vari].sel(lon=lon1, lat=lat1), ds[vari].sel(lon=lon1, lat=lat1)
                 )
 
-    @pytest.mark.parametrize("vectorize", [True, False])
-    def test_wraps(self, tmp_netcdf_filename, vectorize):
+    def test_wraps(self, tmp_netcdf_filename):
         ds = open_dataset(self.nc_file)
 
         # Polygon crosses meridian, a warning should be raised
@@ -680,7 +679,7 @@ class TestSubsetShape:
         self.compare_vals(ds, sub, "tas")
 
         poly = gpd.read_file(self.meridian_multi_geojson)
-        subtas = subset.subset_shape(ds.tas, poly, vectorize=vectorize)
+        subtas = subset.subset_shape(ds.tas, poly)
         np.testing.assert_array_almost_equal(
             float(np.mean(subtas.isel(time=0))), 281.091553
         )
@@ -692,14 +691,13 @@ class TestSubsetShape:
         assert tmp_netcdf_filename.exists()
         with xr.open_dataset(filename_or_obj=tmp_netcdf_filename) as f:
             assert {"tas", "crs"}.issubset(set(f.data_vars))
-            subset.subset_shape(ds, self.meridian_multi_geojson, vectorize=vectorize)
+            subset.subset_shape(ds, self.meridian_multi_geojson)
 
-    @pytest.mark.parametrize("vectorize", [True, False])
-    def test_no_wraps(self, tmp_netcdf_filename, vectorize):
+    def test_no_wraps(self, tmp_netcdf_filename):
         ds = open_dataset(self.nc_file)
 
         with pytest.warns(None) as record:
-            sub = subset.subset_shape(ds, self.poslons_geojson, vectorize=vectorize)
+            sub = subset.subset_shape(ds, self.poslons_geojson)
 
         self.compare_vals(ds, sub, "tas")
 
@@ -724,7 +722,7 @@ class TestSubsetShape:
         assert tmp_netcdf_filename.exists()
         with xr.open_dataset(filename_or_obj=tmp_netcdf_filename) as f:
             assert {"tas", "crs"}.issubset(set(f.data_vars))
-            subset.subset_shape(ds, self.poslons_geojson, vectorize=vectorize)
+            subset.subset_shape(ds, self.poslons_geojson)
 
     def test_all_neglons(self):
         ds = open_dataset(self.nc_file_neglons)

--- a/tox.ini
+++ b/tox.ini
@@ -1,5 +1,5 @@
 [tox]
-envlist = py{36,37,38,39}, py36-slow, py36-nosubset-lm3, py36-xarray, py36-bottleneck, py37-windows, py38-doctest, py38-anaconda, macOS, black, docs, doctests
+envlist = py{36,37,38,39}, py36-slow, py36-nosubset-lm3, py37-xarray, py37-bottleneck, py37-windows, py38-doctest, py38-anaconda, macOS, black, docs, doctests
 requires = pip >= 20.0
 opts = -v
 

--- a/xclim/__init__.py
+++ b/xclim/__init__.py
@@ -6,4 +6,4 @@ from xclim.indicators import ICCLIM, anuclim, atmos, icclim, land, seaIce
 
 __author__ = """Travis Logan"""
 __email__ = "logan.travis@ouranos.ca"
-__version__ = "0.22.2-beta"
+__version__ = "0.22.3-beta"

